File: Problems/872. Leaf-Similar Trees.py
"""
Problem Link: https://leetcode.com/problems/leaf-similar-trees/description/
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        def getTreeLeafs(root, leafs):
            if (root == None):
                return
            if (root.left == None and root.right == None):
                leafs.append(root.val)
                return

            getTreeLeafs(root.left, leafs)
            getTreeLeafs(root.right, leafs)

        leafsRoot1 = []
        getTreeLeafs(root1, leafsRoot1)
        leafsRoot2 = []
        getTreeLeafs(root2, leafsRoot2)

        if (len(leafsRoot1) != len(leafsRoot2)):
            return False

        for i in range(len(leafsRoot1)):
            if (leafsRoot1[i] != leafsRoot2[i]):
                return False

        return True

File: Problems/872. Leaf-Similar Trees.java
/*
Problem Link: https://leetcode.com/problems/leaf-similar-trees/description/
*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

import java.util.*;

class Solution {
    public void dfs(TreeNode root, List<Integer> l) {
        if (root == null)
            return;

        if (root.left == null && root.right == null) {
            l.add(root.val);
            return;
        }

        dfs(root.left, l);
        dfs(root.right, l);
    }

    public boolean leafSimilar(TreeNode root1, TreeNode root2) {
        List<Integer> l1 = new ArrayList<>();
        List<Integer> l2 = new ArrayList<>();

        dfs(root1, l1);
        dfs(root2, l2);

        return l1.equals(l2);
    }
}